include/picongpu/fields/background/templates/twtstight/TWTSTight.tpp

```diff
@@ -0,0 +1,366 @@
+                    return static_cast<float3_X>(NAN);
+            }
+
+            template<typename T_Field>
+            HDINLINE float3_X
+            TWTSTight<T_Field>::operator()(DataSpace<simDim> const& cellIdx, uint32_t const currentStep) const
+            {
+                traits::FieldPosition<fields::YeeCell, T_Field> const fieldPos;
+                return getValue(precisionCast<float_X>(cellIdx), fieldPos(), static_cast<float_X>(currentStep));
+            }
+
+            template<typename T_Field>
+            HDINLINE float3_X TWTSTight<T_Field>::operator()(floatD_X const& cellIdx, float_X const currentStep) const
+            {
+                pmacc::math::Vector<floatD_X, detail::numComponents> zeroShifts;
+                for(uint32_t component = 0; component < detail::numComponents; ++component)
+                    zeroShifts[component] = floatD_X::create(0.0);
+                return getValue(cellIdx, zeroShifts, currentStep);
+            }
+
+            template<typename T_Field>
+            HDINLINE float3_X TWTSTight<T_Field>::getValue(
+                floatD_X const& cellIdx,
+                pmacc::math::Vector<floatD_X, detail::numComponents> const& extraShifts,
+                float_X const currentStep) const
+            {
+                float_64 const time_SI = float_64(currentStep) * dt - tdelay;
+
+                pmacc::math::Vector<floatD_64, detail::numComponents> const fieldPositions_SI
+                    = detail::getFieldPositions_SI(cellIdx, halfSimSize, extraShifts, unit_length, focus_y_SI, phi);
+
+                /* Single TWTS-Pulse */
+                return getTWTSField_Normalized(fieldPositions_SI, time_SI);
+            }
+
+            template<typename T_Field>
+            template<uint32_t T_component>
+            HDINLINE float_X TWTSTight<T_Field>::getComponent(floatD_X const& cellIdx, float_X const currentStep) const
+            {
+                // The optimized way is only implemented for 3D, fall back to full field calculation in 2d
+                if constexpr(simDim == DIM3)
+                {
+                    float_64 const time_SI = float_64(currentStep) * dt - tdelay;
+                    pmacc::math::Vector<floatD_X, detail::numComponents> zeroShifts;
+                    for(uint32_t component = 0; component < detail::numComponents; ++component)
+                        zeroShifts[component] = floatD_X::create(0.0);
+                    pmacc::math::Vector<floatD_64, detail::numComponents> const fieldPositions_SI
+                        = detail::getFieldPositions_SI(cellIdx, halfSimSize, zeroShifts, unit_length, focus_y_SI, phi);
+                    // Explicitly use a 3D vector so that this function compiles for 2D as well
+                    auto const pos = float3_64{
+                        fieldPositions_SI[T_component][0],
+                        fieldPositions_SI[T_component][1],
+                        fieldPositions_SI[T_component][2]};
```

Comment:
Because the compiler would complain in 2D cases if we directly put fieldPositions_SI into the calcTWTSField-functions. In such a 2D case, the compiler nevertheless tries to process the (simDim == DIM3) even when its not fulfilled, finds a type mismatch between floatD and float3 and aborts.